README.md
# Portal Network Tools

[![Join the chat at https://gitter.im/ethereum/eth-portal](https://badges.gitter.im/ethereum/eth-portal.svg)](https://gitter.im/ethereum/eth-portal?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Build Status](https://circleci.com/gh/ethereum/eth-portal.svg?style=shield)](https://circleci.com/gh/ethereum/eth-portal)
[![PyPI version](https://badge.fury.io/py/eth-portal.svg)](https://badge.fury.io/py/eth-portal)
[![Python versions](https://img.shields.io/pypi/pyversions/eth-portal.svg)](https://pypi.python.org/pypi/eth-portal)
[![Docs build](https://readthedocs.org/projects/eth-portal/badge/?version=latest)](http://eth-portal.readthedocs.io/en/latest/?badge=latest)
   

A collection of utilities related to Ethereum's Portal Network

Read more in the [documentation on ReadTheDocs](https://eth-portal.readthedocs.io/). [View the change log](https://eth-portal.readthedocs.io/en/latest/release_notes.html).

## Quickstart

```sh
pip install eth-portal
```

## Developer Setup

If you would like to hack on eth-portal, please check out the [Snake Charmers
Tactical Manual](https://github.com/ethereum/snake-charmers-tactical-manual)
for information on how we do:

- Testing
- Pull Requests
- Code Style
- Documentation

### Development Environment Setup

You can set up your dev environment with:

```sh
git clone git@github.com:ethereum/eth-portal.git
cd eth-portal
virtualenv -p python3 venv
. venv/bin/activate
pip install -e .[dev]
```

### Testing Setup

During development, you might like to have tests run on every file save.

Show flake8 errors on file change:

```sh
# Test flake8
when-changed -v -s -r -1 eth_portal/ tests/ -c "clear; flake8 eth_portal tests && echo 'flake8 success' || echo 'error'"
```

Run multi-process tests in one command, but without color:

```sh
# in the project root:
pytest --numprocesses=4 --looponfail --maxfail=1
# the same thing, succinctly:
pytest -n 4 -f --maxfail=1
```

Run in one thread, with color and desktop notifications:

```sh
cd venv
ptw --onfail "notify-send -t 5000 'Test failure ⚠⚠⚠⚠⚠' 'python 3 test on eth-portal failed'" ../tests ../eth_portal
```

### Release setup

For Debian-like systems:
```
apt install pandoc
```

To release a new version:

```sh
make release bump=$$VERSION_PART_TO_BUMP$$
```

#### How to bumpversion

The version format for this repo is `{major}.{minor}.{patch}` for stable, and
`{major}.{minor}.{patch}-{stage}.{devnum}` for unstable (`stage` can be alpha or beta).

To issue the next version in line, specify which part to bump,
like `make release bump=minor` or `make release bump=devnum`. This is typically done from the
master branch, except when releasing a beta (in which case the beta is released from master,
and the previous stable branch is released from said branch).

If you are in a beta version, `make release bump=stage` will switch to a stable.

To issue an unstable version when the current version is stable, specify the
new version explicitly, like `make release bump="--new-version 4.0.0-alpha.1 devnum"`


docs/release_notes.rst
Release Notes
=============

.. towncrier release notes start

Portal Network Tools v0.2.1 (2022-10-04)
----------------------------------------


Features
~~~~~~~~

- Updated content key encoding ssz scheme & test vectors to new spec, removing the chain id. (`#44 <https://github.com/ethereum/eth-portal/issues/44>`__)


Bugfixes
~~~~~~~~

- Correct the maximum receipt length in SSZ encoding. No specific bug is known to be fixed by this change. (`#40 <https://github.com/ethereum/eth-portal/issues/40>`__)


Improved Documentation
~~~~~~~~~~~~~~~~~~~~~~

- Update ``ln -s`` example in bridge setup guide to use absolute source path. This prevents folks from
  getting a ``Too many levels of symbolic links`` error during setup, because `soft links require an
  absolute source path <https://unix.stackexchange.com/a/180532/251234>`_. (`#42 <https://github.com/ethereum/eth-portal/issues/42>`__)


Portal Network Tools v0.2.0 (2022-09-22)
----------------------------------------


Breaking changes
~~~~~~~~~~~~~~~~

- In order to follow the tip of the chain, you must now specify ``--latest`` or ``-l``.
  See "Run after first installation" under "Run a Bridge Node" (`#38 <https://github.com/ethereum/eth-portal/issues/38>`__)


Features
~~~~~~~~

- Publish data from a historical block range with ``--block-range``. See "Backfill historical blocks" under "Run a Bridge Node" (`#34 <https://github.com/ethereum/eth-portal/issues/34>`__)
- Add the ability to inject specific content items on-demand with ``--content-files``. See "Inject content manually" under "Run a Bridge Node" (`#35 <https://github.com/ethereum/eth-portal/issues/35>`__)


Internal Changes - for Portal Network Tools Contributors
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- When receipts fail their check against the expected root hash, save them to a file in a temporary
  directory (`#32 <https://github.com/ethereum/eth-portal/issues/32>`__)


Miscellaneous changes
~~~~~~~~~~~~~~~~~~~~~

- Merge in latest project template `#41 <https://github.com/ethereum/eth-portal/issues/41>`__


Portal Network Tools v0.1.0 (2022-08-31)
----------------------------------------

Features
~~~~~~~~

- Officially claim support for python versions up to 3.9 (won't support 3.10 until a stable web3
  version does...) (`#29 <https://github.com/ethereum/eth-portal/issues/29>`__)


Improved Documentation
~~~~~~~~~~~~~~~~~~~~~~

- Add a compiled first-time install script to docs (`#25 <https://github.com/ethereum/eth-portal/issues/25>`__)


Internal Changes - for Portal Network Tools Contributors
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - Upgrade ssz to v0.3.0 to get cleaner & faster ssz bytelist decoding
  - Apply ``black`` formatting to setup.py (`#29 <https://github.com/ethereum/eth-portal/issues/29>`__)
- Update project template to get smaller releases, a towncrier duplicate-note fix, CircleCI image
  upgrade, breaking change newsfragments, and pydocstyle failure explanations. (`#30 <https://github.com/ethereum/eth-portal/issues/30>`__)


Miscellaneous changes
~~~~~~~~~~~~~~~~~~~~~

- `#28 <https://github.com/ethereum/eth-portal/issues/28>`__


Breaking changes
~~~~~~~~~~~~~~~~

- Use the portal_historyOffer rpc endpoint of trin, instead of portal_historyStore. This requires a
  more recent version of trin. Follow the `rpc update in trin
  <https://github.com/ethereum/trin/pull/411>`_ for more. (`#27 <https://github.com/ethereum/eth-portal/issues/27>`__)
- Drop Python3.6 to support the latest ssz v0.3.0, which doesn't support py3.6. (`#29 <https://github.com/ethereum/eth-portal/issues/29>`__)


Portal Network Tools v0.1.0-beta.0 (2022-06-02)
-----------------------------------------------

Features
~~~~~~~~

- Launch the bridge with ``python -m eth_portal.bridge``. (`#23 <https://github.com/ethereum/eth-portal/issues/23>`__)
- Add a new script to monitor new headers, with Infura. Also, add utility to encode header fields
  into serialized RLP object. (`#1 <https://github.com/ethereum/eth-portal/issues/1>`__)
- Encode header hash into a Portal History Network content key (`#2 <https://github.com/ethereum/eth-portal/issues/2>`__)
- Encode header content key & value, on each new header hash.  (`#3 <https://github.com/ethereum/eth-portal/issues/3>`__)
- Launch trin nodes with bridge, and push content to them, as new headers arrive. (`#4 <https://github.com/ethereum/eth-portal/issues/4>`__)
- Encode the content key for block receipts (`#7 <https://github.com/ethereum/eth-portal/issues/7>`__)
- Encode receipt values, tested against header's receipt root hash (`#8 <https://github.com/ethereum/eth-portal/issues/8>`__)
- Encode a group of receipts to its Portal Network content value. Currently assumes that we
  switch to using an SSZ list from the RLP list. (`#10 <https://github.com/ethereum/eth-portal/issues/10>`__)
- Push the block's receipts out to trin nodes, after detecting a new header. (`#11 <https://github.com/ethereum/eth-portal/issues/11>`__)
- In bridge launcher, detect if trin "injector" node is already running, then use it. One benefit is
  being able to observe the logs on trin during bridge operation. (`#14 <https://github.com/ethereum/eth-portal/issues/14>`__)
- Encode the content key for block bodies (`#16 <https://github.com/ethereum/eth-portal/issues/16>`__)
- Decode a web3 transaction into a py-evm one, for encoding & hashing (`#17 <https://github.com/ethereum/eth-portal/issues/17>`__)
- Encode uncles and transactions into a block body for Portal Network. (`#18 <https://github.com/ethereum/eth-portal/issues/18>`__)
- Propagate block bodies on the Portal History Network. (`#19 <https://github.com/ethereum/eth-portal/issues/19>`__)


Improved Documentation
~~~~~~~~~~~~~~~~~~~~~~

- Add `bridge` node guide (`#4 <https://github.com/ethereum/eth-portal/issues/4>`__)


Internal Changes - for Portal Network Tools Contributors
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


eth_portal/ssz_sedes.py
from ssz.sedes import ByteList, Container, List, Vector, uint8

#
# History Network Sedes
#

# Content Key serializations
HEADER_TYPE_BYTE = b"\x00"
BODY_TYPE_BYTE = b"\x01"
RECEIPT_TYPE_BYTE = b"\x02"
BLOCK_KEY_SEDES = Container((Vector(uint8, 32),))  # header hash

# Content Value serializations

_MAX_TRANSACTION_LENGTH = 2**24  # ~= 16 million
# Maximum transaction body length is achieved by filling calldata with 0's
# until the block limit of 30M gas is reached.
# At a gas cost of 4 per 0-byte, that produces a 7.5MB transaction. We roughly
# double that size to a maximum of >16 million for some headroom. Note that
# EIP-4488 would put a roughly 1MB limit on transaction length, effectively. So
# increases are not planned (instead, the opposite).

_MAX_TRANSACTION_COUNT = 2**14  # ~= 16k
# 2**14 simple transactions would use up >340 million gas at 21k gas each.
# Current gas limit tops out at 30 million gas.

_MAX_RECEIPT_LENGTH = 2**27  # ~= 134 million
# Maximum receipt length is logging a bunch of data out, currently at a cost of
# 8 gas per byte. Since that is double the cost of 0 calldata bytes, the
# maximum size is roughly half that of the transaction: 3.75 million bytes.
# But there is more reason for protocol devs to constrain the transaction length,
# and it's not clear what the practical limits for receipts are, so we should add more buffer room.
# Imagine the cost drops by 2x and the block gas limit goes up by 8x. So we add 2**4 = 16x buffer.

_MAX_HEADER_LENGTH = 2**13  # = 8192
# Maximum header length is fairly stable at about 500 bytes. It might change at
# the merge, and beyond. Since the length is relatively small, and the future
# of the format is unclear to me, I'm leaving more room for expansion, and
# setting the max at about 8 kilobytes.

_MAX_ENCODED_UNCLES_LENGTH = _MAX_HEADER_LENGTH * 2**4  # = 2**17 ~= 131k
# Maximum number of uncles is currently 2. Using 16 leaves some room for the
# protocol to increase the number of uncles.

_TRANSACTIONS_SEDES = List(
    # Each encoded transaction
    ByteList(_MAX_TRANSACTION_LENGTH),
    _MAX_TRANSACTION_COUNT,
)
# The combined rlp-encoded list of headers:
_UNCLES_SEDES = ByteList(_MAX_ENCODED_UNCLES_LENGTH)
BLOCK_BODY_SEDES = Container((_TRANSACTIONS_SEDES, _UNCLES_SEDES))

BLOCK_RECEIPTS_SEDES = List(
    # Each encoded receipt
    ByteList(_MAX_RECEIPT_LENGTH),
    _MAX_TRANSACTION_COUNT,
)
